File: Algorithms/Sorting/Python/counting_sort.py
def counting_sort(arr):
    """
    Counting Sort implementation in Python.
    Time Complexity: O(n + k) where k is the range of input
    Space Complexity: O(n + k)
    """
    # Find the maximum element in the array
    max_element = max(arr)
    
    # Create a count array to store count of each element
    count = [0] * (max_element + 1)
    
    # Store count of each element in count array
    for num in arr:
        count[num] += 1
    
    # Change count[i] so that count[i] now contains actual
    # position of this element in output array
    for i in range(1, len(count)):
        count[i] += count[i - 1]
    
    # Build the output array
    output = [0] * len(arr)
    for i in range(len(arr) - 1, -1, -1):
        output[count[arr[i]] - 1] = arr[i]
        count[arr[i]] -= 1
    
    return output

# Example usage
if __name__ == "__main__":
    # Test the counting sort
    test_array = [64, 34, 25, 12, 22, 11, 90]
    print("Original array:", test_array)
    sorted_array = counting_sort(test_array)
    print("Sorted array:", sorted_array) 

File: Algorithms/Sorting/C++/counting_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void countingSort(std::vector<int>& arr) {
    int n = arr.size();
    
    // Find the maximum element in the array
    int max_element = *std::max_element(arr.begin(), arr.end());
    
    // Create a count array to store count of each element
    std::vector<int> count(max_element + 1, 0);
    
    // Store count of each element in count array
    for (int num : arr) {
        count[num]++;
    }
    
    // Change count[i] so that count[i] now contains actual
    // position of this element in output array
    for (int i = 1; i <= max_element; i++) {
        count[i] += count[i - 1];
    }
    
    // Build the output array
    std::vector<int> output(n);
    for (int i = n - 1; i >= 0; i--) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }
    
    // Copy the output array to arr[], so that arr[] now
    // contains sorted numbers
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

int main() {
    // Test the counting sort
    std::vector<int> test_array = {64, 34, 25, 12, 22, 11, 90};
    
    std::cout << "Original array: ";
    for (int num : test_array) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    
    countingSort(test_array);
    
    std::cout << "Sorted array: ";
    for (int num : test_array) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    
    return 0;
} 